remove-testing-images/src/index.ts

```diff
@@ -0,0 +1,131 @@
+import * as github from "@actions/github";
+import * as core from "@actions/core";
+
+const token = core.getInput("GITHUB_TOKEN", { required: true });
+const packageName = core.getInput("PACKAGE_NAME", { required: true });
+const tag = core.getInput("TAG");
+
+const octokit = github.getOctokit(token);
+
+await run();
+
+async function run() {
+  const acceptedPackageNames = [
+    "api-testing",
+    "website-testing",
+    "crm-testing",
+    "lld-testing",
+  ];
+
+  if (!acceptedPackageNames.includes(packageName)) {
+    core.setFailed(`package name ${packageName} is not supported.`);
+    return;
+  }
+
+  try {
+    if (tag) {
+      const testingImageId = await findTestingImageByTag();
+
+      if (!testingImageId) {
+        throw new Error(`No testing image found with tag ${tag}`);
+      }
+
+      return RemoveTestingImage(testingImageId);
+    }
+
+    return cleanupUnneededTestingImages();
+  } catch (error) {
+    core.setFailed(error.message);
+  }
+}
+
+async function findTestingImageByTag() {
+  core.info(`🔦 looking for testing image with tag ${tag}...`);
+
+  let imagesWithRequestedTag: number | undefined;
+
+  await octokit.paginate(
+    octokit.rest.packages.getAllPackageVersionsForPackageOwnedByOrg,
+    {
+      package_type: "container",
+      state: "active",
+      package_name: packageName,
+      org: "wunderflats",
+      per_page: 100,
+    },
+    (response, done) => {
+      const testingImages = response.data;
+      const targetImage = testingImages.find((v) =>
+        v.metadata.container.tags.includes(tag)
+      );
+
+      if (targetImage) {
+        imagesWithRequestedTag = targetImage.id;
+        done();
+      }
+
+      return response.data;
+    }
+  );
+
+  return imagesWithRequestedTag;
+}
+
+async function cleanupUnneededTestingImages() {
+  core.info("🔦 finding the unneeded images to cleanup...");
+  const imagesToBeDeleted: number[] = [];
+
+  for await (const response of octokit.paginate.iterator(
```

Comment:
QUESTION, NOT A BLOCKER: I wonder we should have a "limit" on how many images are deleted per run. But I'm not sure if my reasons are much valid:

- Can we get rate-limited esp. during the first runs?
- What if later we introduce a bug and delete too many at a time? (Maybe it's a moot point)
- If the job takes too long during the first runs, it could be marked as failed (but it's not a real problem).

If we _reaaally_ want to, we can add a limit later. And all "consumers" of this action will be updated automatically if we use a reference like `master`.

Reply:
Apparently, we get the rate limit info in the response header, I just tested the remove request, and [here](https://github.com/wunderflats/api/actions/runs/6691836491/job/18179837557) is the result, the limit is on 15K and the images that we have in every repo:
API: 11.5K
Website: almost 6K
LLD: almost 3.5K
CRM: almost 1K

I also tried the paginated fetch request with our current image numbers in API and I didn't get rate limited, so we're fine with all the GitHub APIs.

Reply:
Wow 🙌